src/digest.rs

```diff
@@ -0,0 +1,91 @@
+use libc::{size_t, c_void, c_char};
+use std::ptr;
+
+use crate::yaca_lib as lib;
+use crate::yaca_conv as conv;
+use crate::crypto::Context;
+use crate::*;
+
+
+/// Context for digest operations
+pub struct DigestContext {
+    handle: *mut c_void,
+}
+
+impl Drop for DigestContext {
+    fn drop(&mut self)
+    {
+        unsafe {
+            lib::yaca_context_destroy(self.handle);
+        };
+    }
+}
+
+impl Context for DigestContext {
+    fn get_handle(&self) -> *mut c_void
+    {
+        self.handle
+    }
+}
+
+impl DigestContext {
+    /// Initializes a digest context.
+    pub fn initialize(algo: &DigestAlgorithm) -> Result<DigestContext>
+    {
+        digest_initialize(algo)
+    }
+    /// Feeds the message into the message digest algorithm.
+    pub fn update(&self, message: &[u8]) -> Result<()>
+    {
+        digest_update(self, message)
+    }
+    /// Calculates the final digest.
+    pub fn finalize(&self) -> Result<Vec<u8>>
```

Comment:
You're 100% correct. The context should not be used after finalize. And my initial version of finalize was indeed consuming the context. The problem appeared with GCM and CCM. That's when I reverted this.

Now, how to do what you request of me? I though about this earlier, but did not want to complicate this. The Encrypt/Decrypt context is common for GCM, CCM and all others BCMs. How would you solve this?

Reply:
TL;DR Let's keep it as it is. It mirrors the C api everyone's familiar with. Plus, we'll use this to get a working version. Once that's done, we're going to have to "converge" with the platform api anyway - we'll figure it out then.

----

By no means a request, just an idea for discussion. I'm not sure whether it's preferable to keeping it like it is now. Depends on one's definition of simplicity and crypto common practices.

A convenience generic wrapper function (say, ```finalize_with_tag```) is an option. It could consume the object, return the final block and a tag. It would of course be redundant with respect to the rest of the API - not sure it it's worth adding as a part of the API.

Other than that, there's no way to sensibly do it without overhauling the context hierarchy or introducing associated types. As long as there's one context struct, it can only have a single ```finalize()``` method.

Reply:
ACK

----

Yeah, if we ever do it I'd like to do this properly, with an API that'd be simple for the end use without redundancies. So in case of GCM/CCM user would get a context with only one correct finalize.